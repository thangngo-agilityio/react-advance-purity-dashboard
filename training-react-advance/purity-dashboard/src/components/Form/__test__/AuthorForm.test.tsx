// Libs
import { render, fireEvent, act, waitFor } from '@testing-library/react';

// Components
import { AuthorForm } from '../..';

describe('AuthorForm component', () => {
  const mockProps = {
    onCloseModal: jest.fn(),
    onSubmit: jest.fn(),
  };

  it('should match with snapshot', () => {
    const { container } = render(<AuthorForm {...mockProps} />);

    expect(container).toMatchSnapshot();
  });

  it('should call onSubmit when submit form', async () => {
    const mockNewUserData = {
      name: 'Shoe',
      email: 'shoe@gmail.com',
      avatar: 'asdasd.png',
      role: 'Dev',
      job: 'FE',
      employed: '2024-07-01',
    };

    const { getByLabelText, getByRole } = render(<AuthorForm {...mockProps} />);

    act(() => {
      fireEvent.change(getByLabelText('Name'), {
        target: { value: mockNewUserData.name },
      });
      fireEvent.change(getByLabelText('Email'), {
        target: { value: mockNewUserData.email },
      });
      fireEvent.change(getByLabelText('Avatar'), {
        target: { value: mockNewUserData.avatar },
      });
      fireEvent.change(getByLabelText('Role'), {
        target: { value: mockNewUserData.role },
      });
      fireEvent.change(getByLabelText('Job'), {
        target: { value: mockNewUserData.job },
      });
      fireEvent.change(getByLabelText('Employed'), {
        target: { value: mockNewUserData.employed },
      });
    });

    const saveBtn = getByRole('button', { name: 'Save' });
    fireEvent.click(saveBtn);

    waitFor(() => {
      expect(mockProps.onSubmit).toHaveBeenCalledWith(mockNewUserData, []);
      expect(mockProps.onCloseModal).toHaveBeenCalled();
    });
  });
});
