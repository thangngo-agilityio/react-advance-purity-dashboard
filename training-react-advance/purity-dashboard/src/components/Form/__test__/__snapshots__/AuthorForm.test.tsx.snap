// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthorForm component should match with snapshot 1`] = `
<div>
  <form
    class="chakra-stack css-tl3ftk"
    id="update-product-form"
  >
    <div
      class="chakra-stack css-5crs4f"
    >
      <h2
        class="chakra-heading css-o0bhh9"
      >
        Author
      </h2>
      <div
        class="css-hboir5"
      >
        <div
          class="css-1pjhpfy"
        >
          <div
            class="chakra-form-control css-0"
            role="group"
          >
            <label
              class="chakra-form__label css-byg028"
              for="field-:r0:"
              id="field-:r0:-label"
            >
              Name
            </label>
            <div
              class="chakra-input__group css-bx0blc"
              data-group="true"
            >
              <input
                class="chakra-input css-gstp9h"
                data-testid="edit-field-name"
                id="field-:r0:"
                name="fields.name"
                placeholder="Name"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="css-1pjhpfy"
        >
          <div
            class="chakra-form-control css-0"
            role="group"
          >
            <label
              class="chakra-form__label css-byg028"
              for="field-:r1:"
              id="field-:r1:-label"
            >
              Email
            </label>
            <div
              class="chakra-input__group css-bx0blc"
              data-group="true"
            >
              <input
                class="chakra-input css-gstp9h"
                data-testid="edit-field-email"
                id="field-:r1:"
                name="fields.email"
                placeholder="Email"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="css-1pjhpfy"
      >
        <div
          class="chakra-form-control css-0"
          role="group"
        >
          <label
            class="chakra-form__label css-byg028"
            for="field-:r2:"
            id="field-:r2:-label"
          >
            Avatar
          </label>
          <div
            class="chakra-input__group css-bx0blc"
            data-group="true"
          >
            <input
              accept="image/*"
              class="chakra-input css-gstp9h"
              data-testid="edit-field-avatar"
              id="field-:r2:"
              name="fields.avatar"
              placeholder="Avatar"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="chakra-stack css-5crs4f"
    >
      <h2
        class="chakra-heading css-o0bhh9"
      >
        Function
      </h2>
      <div
        class="css-k008qs"
      >
        <div
          class="css-wgn59o"
        >
          <div
            class="chakra-form-control css-0"
            role="group"
          >
            <label
              class="chakra-form__label css-byg028"
              for="field-:r3:"
              id="field-:r3:-label"
            >
              Role
            </label>
            <div
              class="chakra-input__group css-bx0blc"
              data-group="true"
            >
              <input
                class="chakra-input css-gstp9h"
                data-testid="edit-field-role"
                id="field-:r3:"
                name="fields.role"
                placeholder="Role"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="css-pbhvco"
        >
          <div
            class="chakra-form-control css-0"
            role="group"
          >
            <label
              class="chakra-form__label css-byg028"
              for="field-:r4:"
              id="field-:r4:-label"
            >
              Job
            </label>
            <div
              class="chakra-input__group css-bx0blc"
              data-group="true"
            >
              <input
                class="chakra-input css-gh4tfk"
                id="field-:r4:"
                name="fields.job"
                placeholder="Job"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="chakra-stack css-5crs4f"
    >
      <h2
        class="chakra-heading css-o0bhh9"
      >
        Employed
      </h2>
      <div
        class="css-pbhvco"
      >
        <div
          class="chakra-form-control css-0"
          role="group"
        >
          <label
            class="chakra-form__label css-byg028"
            for="field-:r5:"
            id="field-:r5:-label"
          >
            Employed
          </label>
          <div
            class="chakra-input__group css-bx0blc"
            data-group="true"
          >
            <input
              class="chakra-input css-gh4tfk"
              id="field-:r5:"
              name="fields.employed"
              placeholder="Employed"
              type="date"
              value="Invalid Date"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="css-16xzy9q"
    >
      <button
        class="chakra-button css-2cbh8c"
        data-testid="submit-product-form"
        disabled=""
        form="update-product-form"
        type="submit"
      >
        Save
      </button>
      <button
        class="chakra-button css-42jhew"
        type="button"
      >
        Cancel
      </button>
    </div>
  </form>
</div>
`;
